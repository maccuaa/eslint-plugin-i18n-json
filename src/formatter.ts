import { bold, inverse, underline, white } from "ansi-colors";

import { ESLint } from "@typescript-eslint/utils/dist/ts-eslint";
import { cwd } from "node:process";
import { indentString } from "./util/indent-string";
import { logSymbols } from "./util/log-symbols";
import { plur } from "./util/plur";
import { relative } from "node:path";

const CWD = cwd();

/**
 * Custom eslint formatter for eslint-plugin-i18n-json to allow better error message display.
 * Heavily inspired from https://github.com/sindresorhus/eslint-formatter-pretty.
 */
function formatter(results: ESLint.LintResult[]) {
  let totalErrorsCount = 0;
  let totalWarningsCount = 0;

  const formattedLintMessagesPerFile = results
    .map(({ filePath, messages: fileMessages, errorCount: fileErrorCount, warningCount: fileWarningCount }) => {
      if (fileErrorCount + fileWarningCount === 0) {
        return "";
      }

      totalErrorsCount += fileErrorCount;
      totalWarningsCount += fileWarningCount;

      const relativePath = relative(CWD, filePath);
      const fileMessagesHeader = underline.white(relativePath);

      fileMessages.sort((a, b) => b.severity - a.severity); // display errors first

      const formattedFileMessages = fileMessages
        .map(({ ruleId, severity, message }) => {
          let messageHeader =
            severity === 1
              ? `${logSymbols.warning} ${inverse.yellow(" WARNING ")}`
              : `${logSymbols.error} ${inverse.red(" ERROR ")}`;

          messageHeader += ` ${white(`(${ruleId})`)}`;

          return `\n\n${messageHeader}\n${indentString(message, 2)}`;
        })
        .join("");

      return `${fileMessagesHeader}${formattedFileMessages}`;
    })
    .filter((fileLintMessages) => fileLintMessages.trim().length > 0);

  let aggregateReport = formattedLintMessagesPerFile.join("\n\n");

  // append in total error and warnings count to aggregrate report
  const totalErrorsCountFormatted = `${bold.red(">")} ${logSymbols.error} ${bold.red(
    totalErrorsCount.toString(),
  )} ${bold.red(plur("ERROR", totalErrorsCount))}`;

  const totalWarningsCountFormatted = `${bold.yellow(">")} ${logSymbols.warning} ${bold.yellow(
    totalWarningsCount.toString(),
  )} ${bold.yellow(plur("WARNING", totalWarningsCount))}`;

  aggregateReport += `\n\n${totalErrorsCountFormatted}\n${totalWarningsCountFormatted}`;

  return totalErrorsCount + totalWarningsCount > 0 ? aggregateReport : "";
}

module.exports = formatter;
