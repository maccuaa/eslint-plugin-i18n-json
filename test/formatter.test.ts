import { describe, expect, it } from "vitest";

import { ESLint } from "@typescript-eslint/utils/dist/ts-eslint";
import formatter from "../src/formatter";
import stripAnsi from "./strip-ansi";

const strippedFormatter = (results: ESLint.LintResult[]) => stripAnsi(formatter(results));

describe("formatter", () => {
  it("returns an empty string when there aren't any warnings or errors across all files", () => {
    expect(
      strippedFormatter([
        {
          filePath: "some/file",
          messages: [],
          warningCount: 0,
          errorCount: 0,
          fixableErrorCount: 0,
          fixableWarningCount: 0,
          usedDeprecatedRules: [],
        },
      ])
    ).toMatchSnapshot();
  });

  it("will not display any message for an individual file which does not have any warnings or errors", () => {
    const output = strippedFormatter([
      {
        filePath: "bad/file",
        messages: [
          {
            ruleId: "some-rule",
            severity: 2,
            message: "file is bad",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
        ],
        errorCount: 1,
        warningCount: 0,
        fixableErrorCount: 0,
        fixableWarningCount: 0,
        usedDeprecatedRules: [],
      },
      {
        filePath: "good/file",
        messages: [],
        warningCount: 0,
        errorCount: 0,
        fixableErrorCount: 0,
        fixableWarningCount: 0,
        usedDeprecatedRules: [],
      },
    ]);
    expect(output).toMatchSnapshot();
  });
  it("will display errors before warnings", () => {
    const output = strippedFormatter([
      {
        filePath: "bad/file",
        messages: [
          {
            ruleId: "some-rule",
            severity: 1,
            message: "file has first warning",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
          {
            ruleId: "some-rule",
            severity: 1,
            message: "file has second warning",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
          {
            ruleId: "some-rule",
            severity: 2,
            message: "file is bad",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
          {
            ruleId: "some-rule",
            severity: 2,
            message: "file is pretty bad",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
        ],
        errorCount: 2,
        warningCount: 2,
        fixableErrorCount: 0,
        fixableWarningCount: 0,
        usedDeprecatedRules: [],
      },
    ]);
    expect(output).toMatchSnapshot();
  });
  it("will display issues across many files", () => {
    const output = strippedFormatter([
      {
        filePath: "bad/file",
        messages: [
          {
            ruleId: "some-rule",
            severity: 2,
            message: "file is bad",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
        ],
        errorCount: 1,
        warningCount: 0,
        fixableErrorCount: 0,
        fixableWarningCount: 0,
        usedDeprecatedRules: [],
      },
      {
        filePath: "bad/file2",
        messages: [
          {
            ruleId: "some-rule",
            severity: 1,
            message: "file has a warning",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
        ],
        errorCount: 0,
        warningCount: 1,
        fixableErrorCount: 0,
        fixableWarningCount: 0,
        usedDeprecatedRules: [],
      },
      {
        filePath: "bad/file3",
        messages: [
          {
            ruleId: "some-rule",
            severity: 2,
            message: "file is bad",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
          {
            ruleId: "some-rule",
            severity: 1,
            message: "file has a warning",
            line: 0,
            column: 0,
            endColumn: 0,
            endLine: 0,
            fix: {
              range: [0, 0],
              text: "",
            },
            suggestions: [],
          },
        ],
        errorCount: 1,
        warningCount: 1,
        fixableErrorCount: 0,
        fixableWarningCount: 0,
        usedDeprecatedRules: [],
      },
    ]);
    expect(output).toMatchSnapshot();
  });
});
