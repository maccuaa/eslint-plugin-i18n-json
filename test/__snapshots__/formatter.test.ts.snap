// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`formatter > returns an empty string when there aren't any warnings or errors across all files 1`] = `""`;

exports[`formatter > will display errors before warnings 1`] = `
"bad/file

✖  ERROR  (some-rule)
  file is bad

✖  ERROR  (some-rule)
  file is pretty bad

⚠  WARNING  (some-rule)
  file has first warning

⚠  WARNING  (some-rule)
  file has second warning

> ✖ 2 ERRORS
> ⚠ 2 WARNINGS"
`;

exports[`formatter > will display issues across many files 1`] = `
"bad/file

✖  ERROR  (some-rule)
  file is bad

bad/file2

⚠  WARNING  (some-rule)
  file has a warning

bad/file3

✖  ERROR  (some-rule)
  file is bad

⚠  WARNING  (some-rule)
  file has a warning

> ✖ 2 ERRORS
> ⚠ 2 WARNINGS"
`;

exports[`formatter > will not display any message for an individual file which does not have any warnings or errors 1`] = `
"bad/file

✖  ERROR  (some-rule)
  file is bad

> ✖ 1 ERROR
> ⚠ 0 WARNING"
`;
